pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/FilterOperator.java

```diff
@@ -0,0 +1,91 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.query.runtime.operator;
+
+import java.util.ArrayList;
+import java.util.List;
+import javax.annotation.Nullable;
+import org.apache.pinot.common.utils.DataSchema;
+import org.apache.pinot.core.common.Operator;
+import org.apache.pinot.core.common.datablock.BaseDataBlock;
+import org.apache.pinot.core.common.datablock.DataBlockUtils;
+import org.apache.pinot.core.operator.BaseOperator;
+import org.apache.pinot.query.planner.logical.RexExpression;
+import org.apache.pinot.query.runtime.blocks.TransferableBlock;
+import org.apache.pinot.query.runtime.blocks.TransferableBlockUtils;
+import org.apache.pinot.query.runtime.operator.operands.FilterOperand;
+
+
+public class FilterOperator extends BaseOperator<TransferableBlock> {
+  private static final String EXPLAIN_NAME = "HAVING_FILTER";
```

Comment:
This is giving the impression that it is a hard-coded plan. Currently, the multi-stage engine plan does not support filter in the intermediary stages and wherever possible, filter is pushed down to the leaf stage.

With this PR, we are implementing a FilterOperator that can be used in any of the stages (above leaf), but we are treating it as specifically for executing HAVING filter which may not be the case in future for generic queries.

Reply:
good catch. actually this should just be filter operator. since not only HAVING could generate a intermediary filter node. but also non-equality JOIN clauses like `SELECT * FROM a JOIN b on a.key = b.key WHERE a.val > b.val`.
it would be a rather inefficient plan to ship all equality over though